### ISSUE TITLE
Cannot move out of destructured fixed length vectors in `let` and `match`

### ISSUE THREAD

**Issue by brendanzab** [rust-lang/rust#16418]
``` rust
struct Vector3<S>(S, S, S);

// passes type checking
fn into_vec3_arg<S>([x, y, z]: [S, ..3]) -> Vector3<S> {
    Vector3(x, y, z)
}

fn into_vec3_let<S>(v: [S, ..3]) -> Vector3<S> {
    let [x, y, z] = v; // error: use of moved value: `v[..]`
    Vector3(x, y, z)
}

fn into_vec3_match<S>(v: [S, ..3]) -> Vector3<S> {
    match v {
        // error: use of moved value: `v[..]`
        [x, y, z] => Vector3(x, y, z),
    }
}
```

http://is.gd/bc1U2v

Note that it is impossible to work around this using transmute, because:

``` rust
struct Vector3<S>(S, S, S);

fn into_vec3_transmute<S>(v: [S, ..3]) -> Vector3<S> {
    // error: cannot transmute from a type that contains type parameters
    unsafe { mem::transmute(v) }
}
```

http://is.gd/BeSWOs

**Comment by steveklabnik** [comment]
Updated code sample: https://play.rust-lang.org/?gist=604564384c576d2d3862&version=nightly

**Comment by matthewjasper** [comment]
Fixed in MIR borrowck (and so NLL) cc #23121

**Comment by Mark-Simulacrum** [comment]
Yep, closing as duplicate of https://github.com/rust-lang/rust/issues/26736.
